rewrite-java/src/main/java/org/openrewrite/java/JavaVisitor.java

```diff
@@ -137,7 +137,12 @@ public void maybeAddImport(@Nullable String packageName, String typeName, @Nulla
 
     @Incubating(since = "8.2.0")
     public <S> S service(Class<S> service) {
-        return getCursor().firstEnclosingOrThrow(JavaSourceFile.class).service(service);
+        for (Cursor c = getCursor(); c.getParent() != null; c = c.getParent()) {
+            if (c.getValue() instanceof JavaSourceFile) {
+                return ((JavaSourceFile) c.getValue()).service(service);
+            }
+        }
```

Comment:
Since `service()` needs to traverse to the top of the tree, I wonder if it should leave "breadcrumbs" if the depth exceeds a certain level. E.g. if the depth exceeds 100, it could leave a cached entry at half way up the tree (as a message in the corresponding cursor). The next access would then only have to traverse half the way up (although it would have to check the messages of each cursor). The checking for messages could also be delayed until the depth of 50 has been exceeded.